actionpack/test/controller/parameters/raise_on_unpermitted_params_test.rb

```diff
@@ -24,10 +24,82 @@ def teardown
 
   test "raises on unexpected nested params" do
     params = ActionController::Parameters.new(
-      book: { pages: 65, title: "Green Cats and where to find then." })
+      book: { pages: 65, title: "Green Cats and where to find them." })
 
     assert_raises(ActionController::UnpermittedParameters) do
       params.permit(book: [:pages])
     end
   end
+
+  test "expect never raises on unexpected nested params" do
+    params = ActionController::Parameters.new(
+      book: { pages: 65, title: "Green Cats and where to find them." })
+
+    assert_nothing_raised do
+      params.expect(book: [:pages])
+    end
+  end
+
+  test "expect never raises on unpermitted root params" do
+    params = ActionController::Parameters.new(
+      id: 1,
+      book: { pages: 65, title: "Green Cats and where to find them." })
+
+    assert_nothing_raised do
+      params.expect(book: [:pages, :title])
+      params.expect(:id)
+    end
+  end
+
+  test "expect! raises on unexpected nested params" do
+    params = ActionController::Parameters.new(
+      book: { pages: 65, title: "Green Cats and where to find them." })
+
+    assert_raises(ActionController::UnpermittedParameters) do
+      params.expect!(book: [:pages])
+    end
+  end
+
+  test "expect! raises on unpermitted root params" do
+    params = ActionController::Parameters.new(
+      author_id: 1,
+      id: 1,
+      book: { pages: 65, title: "Green Cats and where to find them." })
+
+    # this one clearly should not raise, everything is covered.
+    assert_nothing_raised do
+      params.expect!(:id, :author_id, book: [:pages, :title])
+    end
+
+    # This one is the tricky one. Should it raise?
+    assert_nothing_raised do
+      params.expect!(book: [:pages, :title])
+    end
+
+    # this one clearly should raise, both a root key and a hash are declared.
+    assert_raises(ActionController::UnpermittedParameters) do
+      params.expect!(:id, book: [:pages, :title])
+    end
+
+    # This one is not quite as obvious. Both scalar root keys are declared
+    assert_raises(ActionController::UnpermittedParameters) do
+      params.expect!(:id, :author_id)
+    end
+
+    assert_raises(ActionController::UnpermittedParameters) do
+      params.expect!(:id)
+    end
+
+    assert_raises(ActionController::UnpermittedParameters) do
+      params[:book].expect!(:pages)
+    end
+
+    assert_nothing_raised do
+      params[:book].expect!(:pages, :title)
+    end
+
+    assert_nothing_raised do
+      params(book: {}).expect!(:pages, :title)
+    end
+  end
```

Comment:
Ok, this test fleshes out the potential behaviors for `expect!`.

I'm guessing that I'm overthinking this again, but this gets quite complex.

The simplest answers:
1. You must specify all keys at all levels.
   This code fails: `User.find(params.expect!(:id)).update(params.expect!(user: [:name]))`
2. Unpermitted root keys are always ignored.
   `params.expect!(:id)` always works, but `params.expect(user: [:name])` raises on `user[age]=` being present but `params[:user].expect(:name)` would not raise.

Reply:
I'm removing all unpermitted parameter checking from expect/!